### PR TITLE
Refactored roulette system

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: #3266, #2887

* **Server Mode**: Both

* **Description of Pull Request**: 
Added a script command to open the roulette window from server side.
Added support for clients >= 2018-05-16
Fixed losing behavior
Thanks to @admkakaroto and @Everade

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -19275,77 +19268,99 @@ void clif_parse_RouletteOpen(int fd, struct map_session_data* sd)
...
-			WFIFOW(fd,8 * count + 8) = i;
-			WFIFOW(fd,8 * count + 10) = j;
-			WFIFOW(fd,8 * count + 12) = rd.nameid[i][j];
-			WFIFOW(fd,8 * count + 14) = rd.qty[i][j];
-			count++;
+	for(int i = 0, offset = 8; i < MAX_ROULETTE_LEVEL; i++) {
+		for(int j = 0; j < MAX_ROULETTE_COLUMNS - i; j++) {
+			WFIFOW(fd,offset + 0) = i;
+			WFIFOW(fd,offset + 2) = j;
+#if PACKETVER < 20180516
```
Comment (sctnightcore) [inline src/map/clif.cpp]: I think it should be` #if PACKETVER < 20180523 ` :thinking:

Comment (Lemongrass3110) [reply src/map/clif.cpp]: No it is correct like this. I just tested it. As you can see on hercules @4144 wrote that he does not know the correct version.

Comment (sctnightcore) [reply src/map/clif.cpp]: :ok_hand:

**Comment by admkakaroto** [comment]
compile

**Comment by ecdarreola** [comment]
Works fine for 2018-05-30! Thanks!

For 2018-06-20, it shows like this
image

No errors though. Sql data already been loaded.

**Comment by Lemongrass3110** [comment]
@ecdarreola which version of 2018-06-20 are you using? I tried with 2018-06-20**e**RagexeRE and it works fine.

grafik

**Comment by ecdarreola** [comment]
@Lemongrass3110 : Was using the Ragexe version. Will that be an issue?

Tested 2018-06-20eRagexe & 2018-06-21aRagexe both. Will test with RagexeRE

**Comment by Lemongrass3110** [comment]
Probably, since they maybe not merge it into the main client yet.

**Comment by admkakaroto** [comment]
when you 'lose' you are not being able to receive the reward:

**Comment by Lemongrass3110** [comment]
Works fine for me. Make sure your inventory has space and everything.

**Comment by Haikenz** [comment]
Works fine for me too. Client 2018-06-21a
https://youtu.be/wCnnjc8r_uk

**Comment by V0rr** [comment]
When you don't have any coins to continue and when you lose but the prize is there to pick, if you click start nothing happens because you don't have any coins or you didn't win to the next level, but after that you can't get the prize anymore.
I fixed this by putting a code after 
`if (!battle_config.feature_roulette) {
		clif_messagecolor(&sd->bl,color_table[COLOR_RED],msg_txt(sd,1497),false,SELF); //Roulette is disabled
		return;
	}`
in clif_parse_roulette_generate:
`if(sd->roulette.claimPrize && sd->roulette_point.bronze <= 0 && sd->roulette_point.silver < 10 && sd->roulette_point.gold < 10 && rd.flag[sd->roulette.stage][sd->roulette.prizeIdx]&1) 
		return;`

**Comment by Leonnix18** [comment]
I tried the roulette system but as I advanced from stage 0 to stage 6, the system only subtracted 1 Bronze Coin. I made this change below to the src/map/clif.ccp and it's working correctly.

I fix by changing this code
---------------------------------
	if( !sd->roulette.stage && sd->roulette_point.bronze <= 0 && sd->roulette_point.silver < 10 && sd->roulette_point.gold < 10 ){
		result = GENERATE_ROULETTE_NO_ENOUGH_POINT;
	}else{
		if (!sd->roulette.stage) {
			if (sd->roulette_point.bronze > 0) {
				sd->roulette_point.bronze -= 1;
				pc_setreg2(sd, ROULETTE_BRONZE_VAR, sd->roulette_point.bronze);
			} else if (sd->roulette_point.silver > 9) {
				sd->roulette_point.silver -= 10;
				sd->roulette.stage = 2;
				pc_setreg2(sd, ROULETTE_SILVER_VAR, sd->roulette_point.silver);
			} else if (sd->roulette_point.gold > 9) {
				sd->roulette_point.gold -= 10;
				sd->roulette.stage = 4;
				pc_setreg2(sd, ROULETTE_GOLD_VAR, sd->roulette_point.gold);
			}
		}

and change it to

	if( (sd->roulette.stage <= 1  && sd->roulette_point.bronze <= 0)? 1:(1 < sd->roulette.stage && sd->roulette.stage <= 4 && sd->roulette_point.silver < 10)? 1: (sd->roulette.stage > 4 && sd->roulette_point.gold < 10)? 1:0 ){
		result = GENERATE_ROULETTE_NO_ENOUGH_POINT;
	}else{
		if (sd->roulette.stage) {
			if (sd->roulette_point.bronze > 0 && sd->roulette.stage <= 1) {
				sd->roulette_point.bronze -= 1;
				pc_setreg2(sd, ROULETTE_BRONZE_VAR, sd->roulette_point.bronze);
			} else if (sd->roulette_point.silver > 9 && sd->roulette.stage > 1 && sd->roulette.stage <= 4) {
				sd->roulette_point.silver -= 10;				
				pc_setreg2(sd, ROULETTE_SILVER_VAR, sd->roulette_point.silver);
			} else if (sd->roulette_point.gold > 9 && sd->roulette.stage > 4) {
				sd->roulette_point.gold -= 10;				
				pc_setreg2(sd, ROULETTE_GOLD_VAR, sd->roulette_point.gold);
			}
		}

**Pull request merged**